File: DAA5.cpp
#include <iostream>
#include <vector>

using namespace std;

// Function to check if it's safe to place a queen at (row, col)
bool isSafe(const vector<vector<int>>& board, int row, int col, int n) {
    // Check this column on the upper side
    for (int i = 0; i < row; i++) {
        if (board[i][col] == 1) {
            return false;
        }
    }

    // Check the upper diagonal on the left side
    for (int i = row, j = col; i >= 0 && j >= 0; i--, j--) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    // Check the upper diagonal on the right side
    for (int i = row, j = col; i >= 0 && j < n; i--, j++) {
        if (board[i][j] == 1) {
            return false;
        }
    }

    return true;
}

// Function to print the board
void printBoard(const vector<vector<int>>& board) {
    cout << "Current board configuration:\n";
    for (const auto& row : board) {
        for (int cell : row) {
            if (cell == 1) {
                cout << "[Q] "; // Queen representation
            } else {
                cout << "[ ] "; // Empty space representation
            }
        }
        cout << endl; // Move to the next line after each row
    }
    cout << endl; // Add an extra line for better separation
}

// Recursive utility function to solve the N-Queens problem
void solveNQueens(vector<vector<int>>& board, int row, int n) {
    // If all queens are placed, print the board
    if (row == n) {
        printBoard(board); // Print the valid board configuration
        return;
    }

    // Try placing queens in all columns of the current row
    for (int col = 0; col < n; col++) {
        if (isSafe(board, row, col, n)) {
            // Place the queen
            board[row][col] = 1;

            // Recursively place the remaining queens
            solveNQueens(board, row + 1, n);

            // Backtrack: remove the queen
            board[row][col] = 0; // Reset the position
        }
    }
}

// Function to initiate the N-Queens problem
void nQueens(int n, int firstRow, int firstCol) {
    vector<vector<int>> board(n, vector<int>(n, 0)); // Create an n x n board initialized to 0

    // Place the first queen
    board[firstRow][firstCol] = 1;

    // Start placing the remaining queens from the next row
    solveNQueens(board, firstRow + 1, n);
}

int main() {
    int n, firstRow, firstCol;

    cout << "Enter the size of the board (N): ";
    cin >> n;

    // Validate board size input
    if (n <= 0) {
        cout << "Invalid board size. Please enter a positive integer." << endl;
        return 1;
    }

    cout << "Enter the row and column to place the first Queen (0-indexed): ";
    cin >> firstRow >> firstCol;

    // Validate the position
    if (firstRow < 0 || firstRow >= n || firstCol < 0 || firstCol >= n) {
        cout << "Invalid position for the first queen. Please enter values between 0 and " << n - 1 << "." << endl;
        return 1;
    }

    cout << "All possible solutions with the first queen placed at (" 
         << firstRow << ", " << firstCol << "):" << endl;
    
    nQueens(n, firstRow, firstCol);

    return 0;
}

File: DAA5.py
import time

def issafe(arr, x, y, n):
    # Check the column for already placed queens
    for row in range(x):
        if arr[row][y] == 1:
            return False
    
    # Check the upper-left diagonal
    row, col = x, y
    while row >= 0 and col >= 0:
        if arr[row][col] == 1:
            return False
        row -= 1
        col -= 1
    
    # Check the upper-right diagonal
    row, col = x, y
    while row >= 0 and col < n:
        if arr[row][col] == 1:
            return False
        row -= 1
        col += 1
    
    return True

def nQueen(arr, x, n, solutions):
    # If all queens are placed
    if x == n:
        temp = []
        for row in range(n):
            temp.append(arr[row][:])  # Copy the current arrangement
        solutions.append(temp)
        return
    
    # Try placing queens in each column of the current row
    for row in range(n):
        if issafe(arr, x, row, n):  # Check if it's safe to place a queen
            arr[x][row] = 1  # Place the queen
            nQueen(arr, x + 1, n, solutions)  # Recur to place the next queen
            arr[x][row] = 0  # Backtrack: remove the queen

def main():
    start_time = time.time()  # Start the timer
    
    n = int(input("Enter number of Queens: "))
    arr = [[0] * n for _ in range(n)]
    solutions = []
        
    # Asking the user to input the position of the first queen
    first_queen_x = int(input(f"Enter row (0 to {n-1}) of the first queen: "))
    first_queen_y = int(input(f"Enter column (0 to {n-1}) of the first queen: "))
        
    # Validate the input coordinates
    if 0 <= first_queen_x < n and 0 <= first_queen_y < n:
        arr[first_queen_x][first_queen_y] = 1  # Place the first queen
        nQueen(arr, first_queen_x + 1, n, solutions)  # Start from the next row
            
        if not solutions:
            print("No solutions found.")
        else:
            print(f"Found {len(solutions)} solutions:")
            for idx, sol in enumerate(solutions, 1):
                print(f"Arrangement {idx}:")
                for row in sol:
                    print(" ".join("Q " if cell == 1 else ". " for cell in row))
                print()
            print("Possible solutions =", idx)
    else:
        print("Invalid position for the first queen. Please try again.")
    
    end_time = time.time()  # End the timer
    execution_time = end_time - start_time  # Calculate execution time
    print(f"Execution time: {execution_time:.6f} seconds")

if __name__ == '__main__':
    main()

# Time complexity = O(N!)